### PR TITLE
[Query rules] Enforce that rule_query requires a platinum license

### PR DESCRIPTION
**Pull request by kderusso**
Enforces that `rule_query` requires a platinum license.

If a user attempts to use the `rule_query` without an adequate license, searches using `rule_query` return the following response: 
```
{
    "error": {
        "root_cause": [
            {
                "type": "security_exception",
                "reason": "current license is non-compliant for [rule_query]",
                "license.expired.feature": "rule_query"
            }
        ],
        "type": "security_exception",
        "reason": "current license is non-compliant for [rule_query]",
        "license.expired.feature": "rule_query"
    },
    "status": 403
}
```

### CODE REVIEW

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/LocalStateEnterpriseSearch.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
+/*
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/LocalStateEnterpriseSearch.java:1]: Note: This class currently isn't used. I had tried to create an extension of the `LocalStateCompositeXPackPlugin` thinking that would be a way to return a non-null license state. I still think it's possible there's a way to do this, so I'm keeping this in for now as it's a draft PR. For now though, I'm relying on a null check in the query builder.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java [hunk 2]
```diff
@@ -283,6 +286,10 @@ protected int doHashCode() {
 
     public static RuleQueryBuilder fromXContent(XContentParser parser) {
         try {
+            XPackLicenseState licenseState = XPackPlugin.getSharedLicenseState();
+            if (licenseState != null && QueryRulesConfig.QUERY_RULES_LICENSE_FEATURE.check(licenseState) == false) {
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java:290]: Null check required for tests to pass, see note below.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java [hunk 3]
```diff
@@ -281,8 +283,11 @@ protected int doHashCode() {
         declareStandardFields(PARSER);
     }
 
-    public static RuleQueryBuilder fromXContent(XContentParser parser) {
+    public static RuleQueryBuilder fromXContent(XContentParser parser, XPackLicenseState licenseState) {
         try {
+            if (QueryRulesConfig.QUERY_RULES_LICENSE_FEATURE.check(licenseState) == false) {
+                throw LicenseUtils.newComplianceException(NAME);
+            }
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java:290]: Should we throw the exception outside of the try block? We don't want to wrap the license exception under the parsing one.

**Review by jimczi — APPROVED** [review]

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/LocalStateEnterpriseSearch.java
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/LocalStateEnterpriseSearch.java:1]: Nice finding! 💯

**Review by carlosdelest — APPROVED** [review]
LGTM, good finding about `LocalStateCompositeXPackPlugin`!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java (hunk 3 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java:290]: Yeah, I'd originally put it there and was trying to make it prettier. 😅 I'll change it back.

**Pull request merged**